examples/structured_data/classification_with_tfdf.py

```diff
@@ -0,0 +1,726 @@
+"""
+
+print(gbt_model.summary())
+
+"""
+## Experiment 2: Decision Forests with target encoding
+
+[Target encoding](https://dl.acm.org/doi/10.1145/507533.507538) is a common preprocessing
+technique for categorical features that convert them into numerical features.
+Using categorical features with high cardinality as-is may lead to overfitting.
+Target encoding aims to replace each categorical feature value with one or more
+numerical values that repesent its co-occerence with the target labels.
+
+More preciely, given a categorical feature, the binary target encoder in this example
+will produce three new numerical features:
+
+1. `positive_frequency`: How many times each feature value occured with a positive target label.
+2. `negative_frequency`: How many times each feature value occured with a negative target label.
+3. `positive_probability`: The probability that the target label is possitive,
+given the feature value, which is computed as
+`positive_frequency / (positive_frequency + negative_frequency)`.
+
+
+Note that target encoding is effective with models that cannot automatically
+learn dense repesentations to categorical features, such as decision forests
+or kernel methods. If neural network models are used, its recommended to
+encode categorical features as embeddings.
+"""
+
+"""
+### Implement Binary Target Encoder
+"""
+
+from keras.engine import base_preprocessing_layer
+
+
+class BinaryTargetEncoding(base_preprocessing_layer.PreprocessingLayer):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        base_preprocessing_layer.keras_kpl_gauge.get_cell("BinaryTargetEncoding").set(
+            True
+        )
+
+    def adapt(self, data):
+        # data is expected to be a integer numpy array ot a Tensor shape [num_exmples, 2].
+        # This contain feature values for a given feature in the dataset, and target values.
+
+        self.compile()
+        self.reset_state()
+
+        # Validate that the data has the expected rank.
+        if data.shape.rank != 2:
+            raise ValueError(f"Data is expected to be of rank 2, got {data.shape.rank}")
+        # Validate the dimensions of the second axis.
+        if data.shape[-1] != 2:
+            raise ValueError(
+                f"The second axis of data is expected to have 2 dimensions, got {data.shape[-1]}"
+            )
+        # Validate that the data has the expected shape.
+        if data.dtype not in [tf.dtypes.int16, tf.dtypes.int32, tf.dtypes.int64]:
+            raise ValueError(
+                f"Data is expected to be a Tensor of type integer, got {data.dtype}"
+            )
+        # Separate the feature values and target values
+        data = self._standardize_inputs(data)
+        feature_values = data[:, 0]
+        target_values = data[:, 1]
+        # Validate that the target has only 0 and 1 values.
+        unique_target_values = tf.sort(tf.unique(target_values).y)
```

Comment:
I think tf.sort and unique are done in memory.
It would be interesting to understand and document what append in this processing layer when used on a large dataset. For example, will the entire dataset be loaded in memory?